tools/cosmovisor/scanner.go

```diff
@@ -109,7 +109,32 @@ func (fw *fileWatcher) CheckUpdate(currentUpgrade upgradetypes.Plan) bool {
 
 	stat, err := os.Stat(fw.filename)
 	if err != nil {
-		// file doesn't exists
+		if os.IsNotExist(err) {
+			return false
+		} else {
+			panic(fmt.Errorf("failed to stat upgrade info file: %w", err))
+		}
+	}
+
+	// check https://github.com/cosmos/cosmos-sdk/issues/21086
+	// If new file is still empty, wait a small amount of time for write to complete
+	if stat.Size() == 0 {
+		for range 10 {
```

Comment:
nit: do you think this could be improved with exponential backoff? I think that this is too arbitrary and might cause again flakyness in the test.

```go
for i range 10 {
...
time.Sleep(math.Pow(2, float64(i)) * time.Millisecond)
```

If it were to fail all 10 times, we would have waited a total of ~2046ms. Let's maybe reduce the amount of retries to 5 (62ms).

Reply:
We can modify in a follow up PR.  In practice the loop never runs more than a few times since the file is quickly finalized.  The failing test was related more to 1) not waiting at all (fixed here) and 2) test case delay being too similar to timeout (already fixed).

In production, retry or waiting some amount of time is necessary to avoid reading empty files (race condition).